birdhouse/deployment/deploy-data-specific-image

```diff
@@ -0,0 +1,88 @@
+#!/bin/bash
+# Deploy data from git repo(s) to local folder(s).
+# This script is run directly on a specific image (such as pavics/crim-jupyter-eo or pavics/crim-jupyter-nlp).
+# It will be used to download and update the different tutorial notebooks associated with a specific image.
+#
+# This is meant to be run on the same host running PAVICS.
+#
+# The data details is specified in a yaml config file (TEMPLATE_CONFIG_YML), using this format :
+# - repo_url: path to the github repo (ex.: https://github.com/crim-ca/pavics-jupyter-images)
+#   branch: name of the branch containing the required version of the data
+#   source_path: path of the desired file or folder in the source repo
+#   dest_dir: directory where the data will be copied to in the specific image where the script is run
+#
+# One or more entries respecting this format can be added to the yaml file, in order to download multiple file/folders.
+#
+# An installation of jq and yq is preferred prior to executing this script.
+#
+# Setting environment variable DEPLOY_DATA_LOGFILE='/path/to/logfile.log'
+# will redirect all STDOUT and STDERR to that logfile so this script will be
+# completely silent.
+
+if [ ! -z "$DEPLOY_DATA_LOGFILE" ]; then
+    exec >>$DEPLOY_DATA_LOGFILE 2>&1
+fi
+
+START_TIME="`date -Isecond`"
+echo "==========
+deploy-data-specific-image START_TIME=$START_TIME"
+
+TEMPLATE_CONFIG_YML="$1"
+if [ -z "$TEMPLATE_CONFIG_YML" ]; then
+    echo "ERROR: missing config.yml file" 1>&2
+    exit 2
+fi
+
+# Empty value could mean typo in the keys in the config file.
+ensure_not_null() {
+    if [ "$*" = null ]; then
+        echo "ERROR: value empty" 1>&2
+        exit 1
+    fi
+}
+
+# Check installation of required package to read yaml files
+if ! command -v jq &> /dev/null; then
+  apt-get -y install jq
+fi
+if ! command -v yq &> /dev/null; then
+  pip install yq
+fi
+
+# Find how many entries are in the config file
+LENGTH="`yq -r '. | length' $TEMPLATE_CONFIG_YML`"
+
+if [ -z $LENGTH ]; then
+    echo "ERROR: empty config file" 1>&2
+    exit 1
+fi
+
+CONFIG_YML="/tmp/notebook_config.yml"
+# Replace environment variables from template to their actual values
+envsubst < $TEMPLATE_CONFIG_YML > $CONFIG_YML
+
+for ((i=0;i<$LENGTH;i++)); do
+  # Extract data from config
+  GIT_REPO="`yq -r .[$i].repo_url $CONFIG_YML`"
+  ensure_not_null "$GIT_REPO"
+
+  BRANCH="`yq -r .[$i].branch $CONFIG_YML`"
+  ensure_not_null "$BRANCH"
+
+  SOURCE_PATH="`yq -r .[$i].source_path $CONFIG_YML`"
+  ensure_not_null "$SOURCE_PATH"
+
+  DEST_DIR=$( eval echo "`yq -r .[$i].dest_dir $CONFIG_YML`")
+  ensure_not_null "$DEST_DIR"
+
+  FULL_URL=$GIT_REPO/branches/$BRANCH/$SOURCE_PATH
+
+  echo "Extracting ${FULL_URL} to ${DEST_DIR}"
+
+  # Download the data from github and copy it to the destination directory
+  svn export --force $FULL_URL $DEST_DIR
```

Comment:
They do have a lot of similarities. I could give it a try and try using the deploy-data script directly, but I am scared of finding a blocking point sometime if we want to add different features to one of the use case.
They are almost similar though, if not the same, for now.

Also, I see that the deploy-data script requires the use of Docker, which means we have to install Docker on the pavics-jupyter-base (this would replace the yq/jq installation actually). I don't know what is the best here, if we want to keep those images to the bare minimum.

@dbyrns I would be curious about your input on this :)

Reply:
I don't have enough information on what @tlvu would like you to reuse to make that call. If he could show us how it could easily be done I'm not against it. As long as we don't have to invest another couple of days to do that. The same apply to the other reuse request, maybe you could talk directly to each other so that it could be done fast.

Reply:
I think the main difference between our scripts is that the `deploy-data` script is meant to run on a generic image that includes Docker and Git, but our new script `deploy-data-specific-image` is meant to be run directly on one of our own specific image, which includes our config.yaml file.

A benefit of using the new script is that it lets us have the yaml file directly on the specific image's repo, keeping it close to its related context. For example, this means, if a developer wants to include a new folder for the crim-eo environment, he just goes to the crim-eo's repo and changes the config there.

I am not sure if we could do this easily with the `deploy-data` script? I think the yaml files are stored directly on the birdhouse-deploy repo? Such as : [deploy_raven_testdata_to_thredds.env](https://github.com/bird-house/birdhouse-deploy/blob/57a320c3583f804838e8c18aeb064266527a0bc9/birdhouse/components/scheduler/deploy_raven_testdata_to_thredds.env) and [deploy-data-raven-testdata-to-thredds.yml](https://github.com/bird-house/birdhouse-deploy/blob/57a320c3583f804838e8c18aeb064266527a0bc9/birdhouse/deployment/deploy-data-raven-testdata-to-thredds.yml)

They are then added as a job in the env.local file : https://github.com/bird-house/birdhouse-deploy/blob/57a320c3583f804838e8c18aeb064266527a0bc9/birdhouse/env.local.example#L177

Reply:
Now that I have more time to digest this change. I would suggest moving this `deploy-data-specific-image` script into the jupyter-pavics-base image so the script and the config yaml (that is in the child image) is part of the same final image.

Why?  So eventually Jenkins can also checkout all the notebooks and run tests on them, like currently with the default Jupyter image.  The e2e repo should not be responsible for checking out the notebooks like currently and will enable that same e2e repo to test against different notebooks depending on the Jupyter image.  That requires implementation of https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/issues/57 but we can start to lay the ground in that direction.

For the scheduler cronjob, this means you don't even have to volume-mount the script from outside since it's already inside the image !

As for re-using the existing generic `deploy-data` script, I think it simply boils down to 
1. being able to use `yq` as `docker run` (currently) or as installed (your case), will have to add a new switch for that.
2. download the repo as a tar/zip archive (your way) instead of `git pull` (current), will have to add a new switch for that as well.
   Note that going the tar/zip archive route nullify the caching provided by `git pull`.  However, `git pull` way waste more space because basically there is a double checkout, one used for caching, one in the `/data/jupyter_user_data/tutorial-notebooks/...`.  Each side have pros and cons so for generic and re-usabilty sake, we can implement both.

In a different PR, once `deploy-data` have the additional mode of operations, the jupyter-pavics-base will wget the script part of the build and make it available inside the image as `deploy-data-specific-image` is, without being directly committed as `deploy-data-specific-image`.